src/sidebar.c

```diff
@@ -253,23 +243,161 @@ void sidebar_update_tag_list(GeanyDocument *doc, gboolean update)
 static gint documents_sort_func(GtkTreeModel *model, GtkTreeIter *iter_a,
 								GtkTreeIter *iter_b, gpointer data)
 {
-	gchar *key_a, *key_b;
 	gchar *name_a, *name_b;
+	GeanyDocument *doc_a, *doc_b;
 	gint cmp;
 
-	gtk_tree_model_get(model, iter_a, DOCUMENTS_SHORTNAME, &name_a, -1);
-	key_a = g_utf8_collate_key_for_filename(name_a, -1);
-	g_free(name_a);
-	gtk_tree_model_get(model, iter_b, DOCUMENTS_SHORTNAME, &name_b, -1);
-	key_b = g_utf8_collate_key_for_filename(name_b, -1);
+	gtk_tree_model_get(model, iter_a, DOCUMENTS_SHORTNAME, &name_a, DOCUMENTS_DOCUMENT, &doc_a, -1);
+	gtk_tree_model_get(model, iter_b, DOCUMENTS_SHORTNAME, &name_b, DOCUMENTS_DOCUMENT, &doc_b, -1);
+
+	/* sort dirs after files (within a directory node) */
+	if (!doc_a && doc_b)
+		cmp = 1;
+	else if (doc_a && !doc_b)
+		cmp = -1;
+	else
+	{
+		gchar *key_a, *key_b;
+		/* g_utf8_collate_key_for_filename() seems to ignore ~. As a result, ~/dev <=> /data
+		 * compares as /dev > /data while ~/dev <=> /etc compares as /dev < /etc. Thus, for sorting
+		 * purposes, treat leading ~ as . which is documented to be special-cased to sort before
+		 * anything else. The side effect, that documents under ~ are always first is actually
+		 * welcome.
+		 */
+		name_a[0] = name_a[0] == '~' ?  '.' : name_a[0];
+		name_b[0] = name_b[0] == '~' ?  '.' : name_b[0];
+		key_a = g_utf8_collate_key_for_filename(name_a, -1);
+		key_b = g_utf8_collate_key_for_filename(name_b, -1);
+		cmp = strcmp(key_a, key_b);
+		g_free(key_b);
+		g_free(key_a);
+	}
+
 	g_free(name_b);
-	cmp = strcmp(key_a, key_b);
-	g_free(key_b);
-	g_free(key_a);
+	g_free(name_a);
 
 	return cmp;
 }
 
+GEANY_EXPORT_SYMBOL
```

Comment:
Why is this necessary?
AFAIU `sidebar_create_store_openfiles()` could be even `static`.

Reply:
Again, for the unit test